### ISSUE TITLE
mips disasm losing xrefs to offsets from $gp

### ISSUE THREAD

**Issue by bowserjklol** [Vector35/binaryninja-api#4839]
**Version and Platform (required):**
- Binary Ninja Version: 3.5.4526
- OS: MacOS
- OS Version: 14.2.1
- CPU Architecture: x64

**Bug Description:**
It appears that the disassembler is losing track of xrefs based on offsets from `$gp` when the value in `$gp` is saved to the stack and later loaded back from the stack.  This worked correctly in 3.4.4271 but no longer functions as expected in both 3.5.4526 (latest stable) and 3.6.4728 (latest dev).

It's of note that In opening other mips binaries in 3.5/3.6, when `$gp` is not manipulated in the function prologue (like it is here in the example program), binja continues to find and label xrefs as I've come to expect.  It _seems_ like this condition only appears when `$gp` gets a new value loaded from the stack (even if that value is the canonical $gp addr).

**Steps To Reproduce:**
1. Open the provided binary and navigate to `http_parser_main`.
2. Observe in 3.4.x, xrefs to `fputs`, `fflush` and others are identified correctly.
3. Observe in 3.5.x and 3.6.x, binja loses the xrefs.

[httpd.zip](https://github.com/Vector35/binaryninja-api/files/13720727/httpd.zip)

**Expected Behavior:**
When the analysis is complete, we should see xrefs to the referenced libc functions.

**Screenshots:**

Example of binja mapping the calls to `fflush` and `fputs` from `http_parser_main` in 3.4.

<img width="740" alt="httpd_3 4 4271" src="https://github.com/Vector35/binaryninja-api/assets/16641949/b1b42bfc-3b01-4b06-bd0a-e89f3f1f0d68">

Example of binja failing to find and label these cross references in 3.5 and 3.6.  Note how it's happy to label the xref for memset but then loses track after `$gp` gets a value from the stack.

_3.5_

<img width="746" alt="httpd_3 5 4526" src="https://github.com/Vector35/binaryninja-api/assets/16641949/e91e724d-eb4f-4194-add7-f9bf91b82792">

_3.6_

<img width="740" alt="httpd_3 6 4728" src="https://github.com/Vector35/binaryninja-api/assets/16641949/c0bb8d8f-0289-4fa7-96d3-387c27f33637">

**Comment by xusheng6** [comment]
@lwerdna could this be related to your recent changes?

**Comment by lwerdna** [comment]
The offending build number is 4469 (4468 works).

**Comment by bowserjklol** [comment]
Just updating the ticket to report 4.0.4911 continues to exhibit the issue.

<img width="779" alt="image" src="https://github.com/Vector35/binaryninja-api/assets/16641949/532c0daa-386a-4279-8fc2-85020efc76c4">

**Comment by psifertex** [comment]
Unfortunately for now we don't have a clean solution to resolving this. The commit that broke this was an important fix in other situations and we don't yet have a good solution for how we can get both working. We did try hard to evaluate if we could fix it before we shipped 4.0 but didn't see any easy solutions. We will definitely re-visit it later but we need to make sure the fix doesn't cause more problems than it fixes!

**Comment by bowserjklol** [comment]
Copy all @psifertex.  All good!  On my end, it's just a matter of keeping some code pinned to the 3.4 build so patterns using BinaryView like `get_symbols_by_name()` then `get_code_refs()` continue to function as expected.

**Comment by bowserjklol** [comment]
Hey all just wanted to bump this issue.  Obviously not as important as the various fixes in 4.0.5336 and 4.0.4958 that have come out since February. Am left wondering if this is something that will be addressed in a future build as the inability correctly map cross references is really degrading my workflow.  Thanks!

**Comment by bowserjklol** [comment]
Pushed up a [small script](https://github.com/bowserjklol/binjaplugins/tree/main/mips-recover-gp-refs) to annotate a bndb (via comments and user-defined data references) as a work around.

This is by no means professional grade and definitely not something anyone should take seriously.  Rather, it merely serves as a (weak) stop-gap for my own triage pipelines.  I can't imagine it'll help anyone else out who is awaiting resolution on this ticket, but wanted to link it just in case.

**Comment by bowserjklol** [comment]
Checking back in on this issue 2 years later (current stable is 5.2.8722).  This issue remains problematic in about 1/3rd of the MIPS binaries I look at and becomes a real problem when triage pipelines start reporting inaccurate results because the API calls they often consume (like `get_code_refs`) don't return complete/accurate results.

**Comment by psifertex** [comment]
I agree -- we really need to resolve this. I'm marking it as high impact and assigning it to this release. No promises that we can get to it still but it will at least make us try to fit it into the schedule and re-evaluate it. Sorry for the delay.

**Comment by bpotchik** [comment]
Fixed in `5.3.9223-dev`.

**Comment by bowserjklol** [comment]
Awesome.  Thanks @bpotchik 🙇
